core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java

```diff
@@ -277,7 +277,7 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
               }
             }
           } catch (IOException e) {
-            throw new RuntimeIOException(e, "Failed to read manifest file: " + manifest.path());
+            throw new UncheckedIOException("Failed to read manifest file: " + manifest.path(), e);
```

Comment:
I do see some public APIs are being affected. E.g  `createTable`.  Do you guys think it is safer to make ` RuntimeOIException` extend `UncheckedIOE` instead of replacing RuntimeIOException completely?

Reply:
+1 for this idea. That would be fewer changes and is a binary compatible change.

Reply:
So would we just make the class extend `UncheckedIOException` and leave everything throwing it as it is now? I personally don't like the way in the arguments it puts the exception first and the message second, I find this confusing as most other Java Exceptions do it the other way around. There is also the "magic" it performs to `String.format` things which again was a bit of a surprise and forced me to look into the internals which I rarely have to do with other Exception types I've used.

You guys know a lot more about how this is used but since its an Unchecked Exception is API compatibility such an issue? Perhaps there is a defined set of classes where it needs to be kept but can be removed everywhere else?

Reply:
Yes, I think we would just make `RuntimeIOException` extend `UncheckedIOException` and leave the current cases as they are, since there isn't a lot of value to removing the class if it extends the more standard one.

I hear your concern about the formatting magic and order of arguments. Here's the context:
1. We want to make it easy to add context into exception messages. It is too common that an exception tells you what happened, but not where it happened. So you might have a `NullPointerException` when writing a record field and not be told which field you need to go look at. The idea behind adding `String.format` into the constructors is to make it as easy as possible to build good exception messages.
2. Because we accept a format string and `Object...` for message formatting, we can't add the wrapped exception at the end of the arg list. The varargs parameter always consumes it. So we had to move it to the beginning.

I though this was a reasonable trade-off for better exception messages. Not sure if other people agree, though.

Reply:
OK, the above context makes sense, but I still think the principle of "least surprises" would lean me to not do this and just make sure people write good messages when they create exceptions. Anyway, at some point I think it would be good to  to have a DEVELOPERS.md (or something equivalent) for getting new developers up to speed with things like the above where the project may differ from what they are used to.

I'll update this PR for `RuntimeIOException` to extend `UncheckedIOException`.